### PR TITLE
Introduce artifact version for terraform

### PR DESCRIPTION
**Pull request by ffjlabo**
**What this PR does / why we need it**:

This PR changes to output artifact versions for terraform development.
Terraform artifact versions are terraform module versions.
Currently, the URLs are not parsed. I'll fix it with another PR.

**Which issue(s) this PR fixes**:

A part of https://github.com/pipe-cd/pipecd/issues/3303

**Does this PR introduce a user-facing change?**:

```release-note
Be able to show artifact versions for terraform deployment
```

### CODE REVIEW

File: pkg/app/piped/cloudprovider/terraform/module.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+	versions := make([]*model.ArtifactVersion, 0, len(modules))
+	for _, m := range modules {
+		versions = append(versions, &model.ArtifactVersion{
+			Kind:    model.ArtifactVersion_TERRAFORM_MODULE,
+			Version: m.Version,
+			Name:    m.Name,
+			Url:     m.Source,
+		})
+	}
```
Comment (knanao) [inline pkg/app/piped/cloudprovider/terraform/module.go:112]: [nits] I think that you needn't to set module var.
```suggestion
	versions := make([]*model.ArtifactVersion, 0, len(tf.Modules))
	for _, m := range tf.Modules {
		versions = append(versions, &model.ArtifactVersion{
			Kind:    model.ArtifactVersion_TERRAFORM_MODULE,
			Version: m.Version,
			Name:    m.Name,
			Url:     m.Source,
		})
	}
```

Comment (ffjlabo) [reply pkg/app/piped/cloudprovider/terraform/module.go:112]: I see, you're right!
Thank you for your suggestion!

File: pkg/app/piped/cloudprovider/terraform/module.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		tfs = append(tfs, &tf)
+	}
+
+	return tfs, nil
+}
+
+// FindArtifactVersions parses artifact versions from Terraform files.
+// For Terraform, module version is an artifact version.
+func FindArtifactVersions(tfs []*File) ([]*model.ArtifactVersion, error) {
+
```
Comment (knanao) [inline pkg/app/piped/cloudprovider/terraform/module.go:99]: [nits] Unnecessary new line.

File: pkg/app/piped/cloudprovider/terraform/module.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// LoadTerraformFiles loads terraform files from a given dir.
+func LoadTerraformFiles(dir string) ([]*File, error) {
+	fileInfos, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	filenames := []string{}
```
Comment (nghialv) [inline pkg/app/piped/cloudprovider/terraform/module.go:48]: nit: just to be unified with our convention.
```suggestion
	filenames := make([]string, 0)
```

File: pkg/app/piped/cloudprovider/terraform/module.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	for _, f := range fileInfos {
+		if f.IsDir() {
+			continue
+		}
+
+		if ext := filepath.Ext(f.Name()); ext != ".tf" {
```
Comment (nghialv) [inline pkg/app/piped/cloudprovider/terraform/module.go:54]: Let's define ".tf" as a constant.

```golang
tfFileExtension = ".tf"
```

File: pkg/app/piped/cloudprovider/terraform/module.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+// Module represents a "module" block in Terraform file.
+type Module struct {
+	Name    string
+	Source  string
+	Version string
+}
+
+// LoadTerraformFiles loads terraform files from a given dir.
+func LoadTerraformFiles(dir string) ([]*File, error) {
```
Comment (nghialv) [inline pkg/app/piped/cloudprovider/terraform/module.go:42]: nit: I think we don't need a slice of pointers.
```suggestion
func LoadTerraformFiles(dir string) ([]File, error) {
```

File: pkg/app/piped/cloudprovider/terraform/module.go (hunk 3 above)
Comment (nghialv) [reply pkg/app/piped/cloudprovider/terraform/module.go:48]: Btw, instead of a list of file names we can store a list of file paths to use later.

**Comment by nghialv** [comment]
Great job! Thank you.
/lgtm

**Comment by knanao** [comment]
Here you go!
/approve

**Pull request merged**